### PR TITLE
[LinalgExt] Use IndexingMapOpInterface for attention

### LINKED ISSUE #21487: [Attention] Attention dispatch memory exceeds max_workgroup_memory_bytes
### What happened?

dispatch:
https://gist.github.com/dan-garvey/e1ed9ad746a4515e6e738fad6e2df2bc

err0r:
```
./dumps/module_decode_bs1$async_dispatch_20.mlir:8:6: error: function 'decode_bs1$async_dispatch_20_attention_DxDxf16_generic' uses 33034304 bytes of shared memory; exceeded the limit of 65536 bytes
      func.func @decode_bs1$async_dispatch_20_attention_DxDxf16_generic() {
     ^
./dumps/module_decode_bs1$async_dispatch_20.mlir:2:2: error: failed to run translation of source executable to target executable for backend #hal.executable.target<"rocm", "rocm-hsaco-fb", {abi = "hip", iree.gpu.target = #iree_gpu.target<arch = "gfx942", features = "", wgp = <compute =  fp64|fp32|fp16|int64|int32|int16|int8, storage =  b64|b32|b16|b8, subgroup =  shuffle|arithmetic, dot =  dp4xi8toi32, mma = [<MFMA_F32_16x16x16_BF16>, <MFMA_F32_32x32x8_BF16>, <MFMA_F32_16x16x32_F8E5M2FNUZ>, <MFMA_F32_16x16x32_F8E5M2FNUZ_F8E4M3FNUZ>, <MFMA_F32_16x16x32_F8E4M3FNUZ>, <MFMA_F32_16x16x32_F8E4M3FNUZ_F8E5M2FNUZ>, <MFMA_F32_32x32x16_F8E5M2FNUZ>, <MFMA_F32_32x32x16_F8E5M2FNUZ_F8E4M3FNUZ>, <MFMA_F32_32x32x16_F8E4M3FNUZ>, <MFMA_F32_32x32x16_F8E4M3FNUZ_F8E5M2FNUZ>, <MFMA_I32_16x16x32_I8>, <MFMA_I32_32x32x16_I8>, <MFMA_F64_16x16x4_F64>, <MFMA_F32_16x16x4_F32>, <MFMA_F32_16x16x16_F16>, <MFMA_F32_32x32x8_F16>], subgroup_size_choices = [64], max_workgroup_sizes = [1024, 1024, 1024], max_thread_count_per_workgroup = 1024, max_workgroup_memory_bytes = 65536, max_workgroup_counts = [2147483647, 2147483647, 2147483647], max_load_instruction_bits = 128, simds_per_wgp = 4, vgpr_space_bits = 16384>>, iree_codegen.default_tuning_spec = #rocm.builtin.tuning_module<"iree_default_tuning_spec_gfx942.mlir">, ukernels = "none"}>
  hal.executable.variant public @rocm_hsaco_fb target(<"rocm", "rocm-hsaco-fb", {abi = "hip", iree.gpu.target = #iree_gpu.target<arch = "gfx942", features = "", wgp = <compute =  fp64|fp32|fp16|int64|int32|int16|int8, storage =  b64|b32|b16|b8, subgroup =  shuffle|arithmetic, dot =  dp4xi8toi32, mma = [<MFMA_F32_16x16x16_BF16>, <MFMA_F32_32x32x8_BF16>, <MFMA_F32_16x16x32_F8E5M2FNUZ>, <MFMA_F32_16x16x32_F8E5M2FNUZ_F8E4M3FNUZ>, <MFMA_F32_16x16x32_F8E4M3FNUZ>, <MFMA_F32_16x16x32_F8E4M3FNUZ_F8E5M2FNUZ>, <MFMA_F32_32x32x16_F8E5M2FNUZ>, <MFMA_F32_32x32x16_F8E5M2FNUZ_F8E4M3FNUZ>, <MFMA_F32_32x32x16_F8E4M3FNUZ>, <MFMA_F32_32x32x16_F8E4M3FNUZ_F8E5M2FNUZ>, <MFMA_I32_16x16x32_I8>, <MFMA_I32_32x32x16_I8>, <MFMA_F64_16x16x4_F64>, <MFMA_F32_16x16x4_F32>, <MFMA_F32_16x16x16_F16>, <MFMA_F32_32x32x8_F16>], subgroup_size_choices = [64], max_workgroup_sizes = [1024, 1024, 1024], max_thread_count_per_workgroup = 1024, max_workgroup_memory_bytes = 65536, max_workgroup_counts = [2147483647, 2147483647, 2147483647], max_load_instruction_bits = 128, simds_per_wgp = 4, vgpr_space_bits = 16384>>, iree_codegen.default_tuning_spec = #rocm.builtin.tuning_module<"iree_default_tuning_spec_gfx942.mlir">, ukernels = "none"}>) {
 ^
```

Additionally, takes ages to compile. I recently had a similar issue when attention was being decomposed at the torch level, but this is using the custom sharktank kernel. 

I'm not certain how to debug this, so I'd appreciate any help. It may be worth noting that without the following python code, the graph compiles just fine:
```
def sliding_window_only_mask(seq_len, window_left, window_right):
    i = torch.arange(seq_len).unsqueeze(1)  # [T, 1]
    j = torch.arange(seq_len).unsqueeze(0)  # [1, T]
    # Create mask where j is outside [i - left, i + right]
    valid = (j >= (i - window_left)) & (j <= (i + window_right))
    mask = torch.where(valid, torch.tensor(0.0, dtype=torch.float16), torch.tensor(float('-inf'), dtype=torch.float16))
    return mask

if self.attention_window[0] != -1:
    seq_len = xk.size(-2) 
    window_mask = sliding_window_only_mask(seq_len, self.attention_window[0], self.attention_window[1])
    attention_mask = torch.minimum(attention_mask, window_mask)
```
(This enforces a 1024 window on the attention mask, s.t. if you are computing the token at seq_len 1032, it will ignore the first 8 tokens)

### What component(s) does this issue relate to?

_No response_

### Version information

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by IanWood1**
Use `IndexingMapOpInterface` for attention which allows using the upstream function to drop the unit dims. 

- Added pattern to drop all attention unit dims & erased old one that only dropped dims on the mask.
- `getStaticLoopRanges` return type was modified to match the `IndexingMapOpInterface`
- Fix `KernelConfig.cpp` to handle non-existent M dims.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp [hunk 1]
```diff
@@ -1340,21 +1340,26 @@ static LogicalResult setAttentionIntrinsicBasedVectorDistributionConfig(
           op.getQueryMap(), op.getKeyMap(), op.getValueMap(), op.getOutputMap())
           .value();
 
-  int64_t mDim = opInfo.getMDims().back();
+  std::optional<int64_t> mDim;
```
Comment (IanWood1) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp]: @Groverkss is same thing needed for the other dimensions? I wasn't sure exactly how to handle it if the others don't exist.

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp]: I think if other dims dont exist you could treat them as unit dimension.

**Review by kuhar — COMMENTED** [review]
@bangtianliu @Max191 do we have to update the attention tuner after this lands?

**Comment by Max191** [comment]
> @bangtianliu @Max191 do we have to update the attention tuner after this lands?

We might need to replace some asserts with graceful failures (i.e., when not all dimensions are present), but otherwise I think there shouldn't be any big changes.

**Comment by bangtianliu** [comment]
> @bangtianliu @Max191 do we have to update the attention tuner after this lands?

Yeah, tuner should not be affected but I can check after this PR is landed.

**Review by hanhanW — COMMENTED** [review]
> Fix KernelConfig.cpp to handle non-existent M dims.

The implementation looks good to me, can you add a test for this part? I'm not familiar with the structure, it looks like you may add the test to [config_vector_distribute_gfx942.mlir](https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/config_vector_distribute_gfx942.mlir)

**Comment by IanWood1** [comment]
I added a test but I don't know enough about attention's lowering to check for anything but no crash.

**Comment by hanhanW** [comment]
> I added a test but I don't know enough about attention's lowering to check for anything but no crash.

I think it is good enough for now, as it reflects the change. If we really need (or want) to have e2e tests, we can consider updating https://github.com/iree-org/iree/blob/main/tests/e2e/attention/generate_e2e_attention_tests.py

I don't have enough context, so I'm not going to make the suggestion.

**Review by hanhanW — APPROVED** [review]
